site/src/components/DurationField/DurationField.tsx

```diff
@@ -0,0 +1,152 @@
+  // especially when a user clears the input field.
+  durationFieldValue: string;
+};
+
+export const DurationField: FC<DurationFieldProps> = (props) => {
+  const { value: parentValue, onChange, helperText, ...textFieldProps } = props;
+  const [state, setState] = useState<State>(() => initState(parentValue));
+  const currentDurationInMs = durationInMs(
+    state.durationFieldValue,
+    state.unit,
+  );
+
+  useEffect(() => {
+    if (parentValue !== currentDurationInMs) {
+      setState(initState(parentValue));
+    }
+  }, [currentDurationInMs, parentValue]);
+
+  return (
+    <div>
+      <div
+        css={{
+          display: "flex",
+          gap: 8,
+        }}
+      >
+        <TextField
+          {...textFieldProps}
+          type="number"
+          css={{ maxWidth: 160 }}
+          value={state.durationFieldValue}
+          onChange={(e) => {
+            const durationFieldValue = e.currentTarget.value;
+
+            setState((state) => ({
+              ...state,
+              durationFieldValue,
+            }));
+
+            const newDurationInMs = durationInMs(
+              durationFieldValue,
+              state.unit,
+            );
+            if (newDurationInMs !== parentValue) {
+              onChange(newDurationInMs);
+            }
+          }}
+          inputProps={{
+            step: 1,
+          }}
+        />
+        <Select
+          disabled={props.disabled}
+          css={{ width: 120, "& .MuiSelect-icon": { padding: 2 } }}
+          value={state.unit}
+          onChange={(e) => {
+            const unit = e.target.value as TimeUnit;
+            setState(() => ({
+              unit,
+              durationFieldValue:
+                unit === "hours"
+                  ? durationInHours(currentDurationInMs).toString()
+                  : durationInDays(currentDurationInMs).toString(),
+            }));
+          }}
+          inputProps={{ "aria-label": "Time unit" }}
+          IconComponent={KeyboardArrowDown}
+        >
+          <MenuItem
+            value="hours"
+            disabled={!canConvertDurationToHours(currentDurationInMs)}
+          >
+            Hours
+          </MenuItem>
+          <MenuItem
+            value="days"
+            disabled={!canConvertDurationToDays(currentDurationInMs)}
+          >
+            Days
+          </MenuItem>
+        </Select>
+      </div>
+
+      {helperText && (
+        <FormHelperText error={props.error}>{helperText}</FormHelperText>
+      )}
+    </div>
+  );
+};
+
+function initState(value: number): State {
+  const unit = suggestedTimeUnit(value);
+  const durationFieldValue =
+    unit === "hours"
+      ? durationInHours(value).toString()
+      : durationInDays(value).toString();
+
+  return {
+    unit,
+    durationFieldValue,
+  };
+}
+
+function durationInMs(durationFieldValue: string, unit: TimeUnit): number {
+  const durationInMs = parseInt(durationFieldValue);
```

Comment:
Just to be extra, super safe, can you add a radix argument of 10? `parseInt` automatically tries to figure out what radix you want if you don't give it one, but sometimes it chooses something that isn't base-10:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#parameters

Reply:
I'm wondering how you hold so much knowledge about web platforms. 🧠